### ISSUE TITLE
Server Issues

### ISSUE THREAD

**Issue by NielsPilgaard** [WesCook/Nutrition#139]
Hello! One of my users reported a strange issue with Nutrition on his server.
TL;DR: Nutrition buffs stop being renewed when a 3rd player joins the server.

### Original issue
OP: @Kortako
**Issue Description**

I'm playing on my own private server with 2 other friends, and we each have fairly high (>90%) nutrition values for all of our nutrition categories.  When only 2 of us are logged in, all is fine, but if the third logs in what will often happen is the Nutrition buffs will expire for some of us, and we'll lose any extra hearts granted by that except for the ones you get for eating 10 types of food.  If one of us logs out, then the nutrition starts "recalculating" and we get our buffs/hearts back.

This happens fairly often when all of us are online and even a server restart won't fix this.  It seems to only happen when the third player joins, and it doesn't seem to be any player in particular causing it.  I'm pretty sure this only happens on a server as far as I'm aware, and I've noticed at least one other person mentioning the same issue on the Discord support channel (for Enigmatica 2 Expert).

**Mod version**

Nutrition-1.12.2-4.0.0

**Screenshots**

I can add screenshots if needed, but I need to wait until all of us are on the server at the same time for it to happen again.

**Comment by WesCook** [comment]
That's... really weird.  There isn't any code that looks at the number of players online.  If there's been multiple reports of that behaviour though then it sounds like a serious bug.

Thanks for forwarding that report to me.

**Comment by NielsPilgaard** [comment]
I was just told that this bug also temporarily removes hearts gained from Spice of Life: Carrot Edition, making me think the bug is not on your side :)

**Comment by NielsPilgaard** [comment]
Edit: The reporter remembered wrong, only Nutrition effects are affected by this bug, reopening.

**Comment by WesCook** [comment]
That would have been too easy, wouldn't it?

**Comment by Kortako** [comment]
Just chiming in here, I'm the one who originally reported this, and I did mention that I did keep the hearts gained from eating food, which I'm assuming is the Spice of Life hearts.  I wasn't sure if they were related or not.

@WesCook, if there's anything I can do to help debug or give more info, just let me know.

**Comment by Allalous** [comment]
Me and my group of friends (9 of us in total) play on a pack that my friend made and we noticed a week ago that whenever an odd number person joins to play, everyone's nutrition glitches from whatever they are to having zero of the added hearts, no buffs, no nothing until the players online go to an even number. We have tried everything we can think of to fix it (we tried the /nutrition reload command) and we don't have Spice of Life

**Comment by WesCook** [comment]
That's extremely helpful.  Thank you, @Allalous.  I suspect I may have introduced a bug in the latest refactor ([v4.0.0](https://github.com/WesCook/Nutrition/releases)).

**Comment by WesCook** [comment]
Just to update this, I have been able to replicate this bug in a local environment.  That's good!

**Comment by fajcyk** [comment]
I don't know if this is going to help, but I'm also affected by this bug (playing Enigmatica 2 Expert 1.54), but I lost my buffs only after dying for the first time since updating the mod.

**Comment by WesCook** [comment]
Sorry for the wait everyone.  I've put out a fix for this issue now with [v4.1.0](https://github.com/WesCook/Nutrition/releases/tag/v4.1.0).

To those affected, would you mind testing just to make sure the bug is fixed?  I found that three players were required on a server before the issue would occur (my mistake was testing with only two before now).  It works well in my testing now, but I'd appreciate the peace of mind from another confirmation!

Thank you to all those who reported this one.

**Comment by NielsPilgaard** [comment]
Nice work, thank you! I'll make sure to include the update.

**Comment by WesCook** [comment]
Awesome, thanks.  I'll probably be pushing a v4.2 before too long once I clean up some of the newly requested mod support, but I just wanted to get a bugfix out quickly for this issue.

If you've customized the jsons to your pack then it won't affect you anyway though.

**Comment by WesCook** [comment]
As promised, [v4.2.0](https://github.com/WesCook/Nutrition/releases/tag/v4.2.0) is now out.

I will now stop spamming this issue.
